### ISSUE TITLE
llvm failed to generate code when trait_upcasting feature used

### ISSUE THREAD

**Issue by mikialex** [rust-lang/rust#99551]
I tried this code:

```rust
#![feature(trait_upcasting)]

pub trait A {}
pub trait B {}

pub trait C: A + B {}
impl<X: A + B> C for X {}

pub fn test<'a, T>(view: T) -> Option<&'a mut dyn B>
where
  T: IntoIterator<Item = &'a mut dyn B>,
{
  return Some(view.into_iter().next().unwrap());
}

fn main() {
  let mut a: Vec<Box<dyn C>> = Vec::new();
  test(a.iter_mut().map(|c| c.as_mut() as &mut dyn B));
}


```

I expected to see this happen: compile successfully

Instead, this happened: get a compile error:

```
Invalid InsertValueInst operands!
  %13 = insertvalue { {}*, [3 x i64]* } %12, i8* %11, 1, !dbg !170
in function _ZN6viewer6viewer10controller4test1t28_$u7b$$u7b$closure$u7d$$u7d$17h1000438316d202ffE
LLVM ERROR: Broken function found, compilation aborted!
```

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (87588a2af 2022-07-13)
binary: rustc
commit-hash: 87588a2afd9ca903366f0deaf84d805f34469384
commit-date: 2022-07-13
host: x86_64-apple-darwin
release: 1.64.0-nightly
LLVM version: 14.0.6
```

no other backtrace logged

only debug build has this issue

**Comment by crlf0710** [comment]
Seems no longer panicks on today's playground: 1.65.0-nightly
(2022-08-16 86c6ebee8fa0a5ad1e18)

Is it worth bisecting what has changed in this month?

**Comment by crlf0710** [comment]
Did a roughly bisect on the pre-built nightly compilers:
```text
fetching https://static.rust-lang.org/dist/2022-08-12/channel-rust-nightly-git-commit-hash.txt
nightly manifest 2022-08-12: 40 B / 40 B [======================================================] 100.00 % 100.29 KB/s converted 2022-08-12 to 20ffea6938b5839c390252e07940b99e3b6a889a

fetching https://static.rust-lang.org/dist/2022-08-13/channel-rust-nightly-git-commit-hash.txt
nightly manifest 2022-08-13: 40 B / 40 B [======================================================] 100.00 % 117.20 KB/s converted 2022-08-13 to f22819bcce4abaff7d1246a56eec493418f9f4ee
```
Seems it got fixed between 20ffea6938b5839c390252e07940b99e3b6a889a..f22819bcce4abaff7d1246a56eec493418f9f4ee range. And the llvm 15 update is between them.
